src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -133,26 +163,41 @@ private sealed class Context : IVsFreeThreadedFileChangeEvents2, IContext
             private readonly object _gate = new object();
             private bool _disposed = false;
             private readonly HashSet<FileWatchingToken> _activeFileWatchingTokens = new HashSet<FileWatchingToken>();
-            private uint _directoryWatchCookie;
 
-            public Context(FileChangeWatcher fileChangeWatcher, string? directoryFilePath)
+            /// <summary>
+            /// The list of cookies we used to make watchers for <see cref="_watchedDirectories"/>.
+            /// </summary>
+            /// <remarks>
+            /// This does not need to be used under <see cref="_gate"/>, as it's only used inside the actual queue of file watcher
+            /// actions.
+            /// </remarks>
+            private List<uint> _directoryWatchCookies = new List<uint>();
+
+            public Context(FileChangeWatcher fileChangeWatcher, ImmutableArray<WatchedDirectory> watchedDirectories)
             {
                 _fileChangeWatcher = fileChangeWatcher;
+                _watchedDirectories = watchedDirectories;
                 _noOpFileWatchingToken = new FileWatchingToken();
 
-                if (directoryFilePath != null)
+                foreach (var watchedDirectory in watchedDirectories)
                 {
-                    if (!directoryFilePath.EndsWith("\\"))
-                    {
-                        directoryFilePath += "\\";
-                    }
-
-                    _directoryFilePath = directoryFilePath;
-
                     _fileChangeWatcher.EnqueueWork(
                         async service =>
                         {
-                            _directoryWatchCookie = await service.AdviseDirChangeAsync(_directoryFilePath, watchSubdirectories: true, this).ConfigureAwait(false);
+                            var cookie = await service.AdviseDirChangeAsync(watchedDirectory.Path, watchSubdirectories: true, this).ConfigureAwait(false);
+                            _directoryWatchCookies.Add(cookie);
+
+                            if (watchedDirectory.ExtensionFilter != null)
+                            {
+                                // TODO: switch to proper reference assemblies
+                                var filterDirectoryChangesAsyncMethod = service.GetType().GetMethod("FilterDirectoryChangesAsync");
```

Comment:
This will be fixed up in master-vs-deps.

Reply:
> we don't have 16.7 images yet (those come tomorrow)

I'm willing to wait until tomorrow for you to not use reflection.

Reply:
I have low confidence it'll work on the images we're getting since those were Preview 1 images and I think the API changes we needed went into Preview 2. Putting this in master and then removing the reflection in master-vs-deps is also helpful because it limits the code skew between the two branches.

Reply:
open tracking issue to get us off reflectio nhere?

Reply:
I'm just going to make the PR later today